frontend/src/metabase/query_builder/components/expressions/ExpressionEditorTextfield/ExpressionEditorTextfield.jsx

```diff
@@ -320,14 +320,18 @@ class ExpressionEditorTextfield extends React.Component {
     }
   };
 
-  handleExpressionChange(source) {
+  handleExpressionChange = source => {
+    if (source !== this.state.source) {
+      this.setState({ hasChanges: true });
```

Comment:
Could you explain why you needed to track `hasChanges` and show the error message only when it has changed?